lib/Pipes/TaggedFunctionKind.cpp

```diff
@@ -49,18 +55,223 @@ TaggedFunctionKind::symbolToTarget(const llvm::Function &Symbol) const {
+
+  std::set<std::string> Targets;
+  const auto &IDF = Model->ImportedDynamicFunctions;
 
-  if (not Ctx.containsReadOnlyContainer(BinaryCrossRelationsRole))
+  for (const Change<Binary> &Change : ModelDiff->Changes) {
+    revng_assert(not Change.Path.empty());
+
+    // A new function has been added or removed?
+    if (pathAsString<Binary>(Change.Path) == "/Functions") {
+      if (Change.New) {
+        // Invalidate all targets
+        for (const model::Function &Function : Model->Functions)
+          Targets.insert(Function.Entry.toString());
+      } else if (Change.Old) {
+        // Invalidate the function and its callers
+        auto Function = std::get<model::Function>(*Change.Old);
+        Targets.insert(Function.Entry.toString());
+        insertCallers(CrossRelations, Function.Entry, Targets);
+      }
+    }
+
+    // A new prototype has been added or removed?
+    if (pathAsString<Binary>(Change.Path) == "/Types") {
+      if (Change.New) {
+        // Invalidate all targets
+        for (const model::Function &Function : Model->Functions)
+          Targets.insert(Function.Entry.toString());
+      } else if (Change.Old) {
+        auto Prototype = std::get<model::TypePath>(*Change.Old);
+
+        for (const model::Function &Function : Model->Functions) {
+          if (Prototype == Function.Prototype) {
+            // Invalidate the function and its callers
+            Targets.insert(Function.Entry.toString());
+            insertCallers(CrossRelations, Function.Entry, Targets);
+          }
+        }
+
+        for (const model::DynamicFunction &Function : IDF) {
+          if (Prototype == Function.Prototype) {
+            // Invalidate callers of the model::DynamicFunction
+            insertCallers(CrossRelations, Function.OriginalName, Targets);
+          }
+        }
+      }
+    }
+
+    // Attributes have changes?
+    auto MaybeChangedAttributes = AttrMatcher.match<MetaAddress>(Change.Path);
+    if (MaybeChangedAttributes) {
+      auto Entry = std::get<MetaAddress>(*MaybeChangedAttributes);
+      using FunctionAttribute = model::FunctionAttribute::Values;
+      bool NewAttribute = std::get<FunctionAttribute>(*Change.New);
+      if (NewAttribute == FunctionAttribute::NoReturn) {
+        // Invalidate callers
+        insertCallers(CrossRelations, Entry, Targets);
+      }
+    }
+
+    // A new dynamic function has been added or removed?
+    if (pathAsString<Binary>(Change.Path) == "/ImportedDynamicFunctions") {
+      if (Change.New) {
+        // Invalidate all targets
+        for (const model::Function &Function : Model->Functions)
+          Targets.insert(Function.Entry.toString());
+      } else if (Change.Old) {
+        // Invalidate callers of the model::DynamicFunction
+        for (const model::DynamicFunction &Function : IDF)
+          insertCallers(CrossRelations, Function.OriginalName, Targets);
+      }
+    }
+  }
```

Comment:
I think we want to do this for regular functions too.
You probably got it mixed up with the types and functions above.

Reply:
We invalidate all targets if we have an addition, or invalidate the function itself (not needed for if we have a `model::DynamicFunction` & its callers. This is what is happening currently both for `model::Function` and `model::DynamicFunction`. I didn't understand what is making you perplexed.

Reply:
OK, put them close together, they are almost the same code.

Reply:
Sure.